onnxruntime/core/session/inference_session_utils.cc

```diff
@@ -0,0 +1,211 @@
+      session_options.graph_optimization_level = TransformerLevel::Level3;
+      return Status::OK();
+
+    default:
+      LOGS(logger, ERROR) << "Unsupported graph_optimization_level value in ORT config: " << value;
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Unsupported graph_optimization_level value in ORT config: ", value);
+  }
+}
+
+static Status SetEnableProfiling(SessionOptions& session_options,
+                                 int value,
+                                 const logging::Logger& logger) {
+  if (value != 0 && value != 1) {
+    LOGS(logger, ERROR) << "Unsupported value for enable_profiling option: " << value;
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Unsupported value for enable_profiling option: ", value);
+  }
+
+  LOGS(logger, INFO) << "Setting enable_profiling to " << (value == 0 ? "false" : "true");
+  session_options.enable_profiling = (value == 0 ? false : true);
+  return Status::OK();
+}
+
+//---------------------------------------------------
+//--- end of session options related helpers ---
+//---------------------------------------------------
+
+//---------------------
+//--- end of local helpers ---
+//---------------------
+
+Status InferenceSessionUtils::ParseOrtConfigJsonInModelProto(const ONNX_NAMESPACE::ModelProto& model_proto) {
+  if (is_model_checked_for_ort_config_json_) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "The Model Proto has already been checked for the ORT config json.");
+  }
+
+  for (const auto& metadata_field : model_proto.metadata_props()) {
+    if (metadata_field.has_key() && metadata_field.key() == inference_session_utils::ort_config_key) {
+      LOGS(logger_, INFO)
+          << "Found session/run/environment configuration in the model file to be used while running the model";
+
+      try {
+        parsed_json_ = json::parse(metadata_field.value());
+        // set the flag indicating that the model has the ORT config json.
+        is_json_available_ = true;
+      } catch (const std::exception& e) {
+        LOGS(logger_, ERROR) << "Json stored in the `ort_config` key cannot be parsed. Error message: " << e.what();
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Json stored in the `ort_config` key cannot be parsed. Error message: ", e.what());
+      }
+
+      break;
+    }
+  }
+
+  // all steps complete, set the flag indicating that the model has been checked for the ORT config json.
+  is_model_checked_for_ort_config_json_ = true;
+  return Status::OK();
+}
+
+Status InferenceSessionUtils::ParseSessionOptionsFromModelProto(SessionOptions& session_options) {
+  if (!is_model_checked_for_ort_config_json_) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "The Model Proto hasn't been checked for the ORT config json.");
+  }
+
+  if (!is_json_available_ || parsed_json_.contains(inference_session_utils::session_options_key)) {
```

Comment:
If the user has explicitly indicated to read the configs from the model, is_json_available *must* be true or else it's an error, right? Same goes for session options key.

Reply:
I feel - Not necessarily. User could indicate a *preference* to look into the model for config options, but all models need not necessarily have configs in which case, we log the absence of the config (I ll check if this logging is being done) and use the default session options.

The following is a very valid scenario to me - 
Someone has hard-coded their scoring script in such a way that they ask to ORT to ALWAYS look into the model for config options, but the model itself might be a variable for the scoring script and not all models need to have ort config in them....

Reply:
Per offline discussion, returning a status indicating failure if the model doesn't contain a json and if the env var for loading config from model is set to true.